bionic/aip/docker_image_builder.py

```diff
@@ -0,0 +1,144 @@
+"""
+Builds a docker image for Google AI Platform execution using the current Python
+environment.
+"""
+import pathlib
+import subprocess
+from concurrent.futures import Future
+from concurrent.futures.thread import ThreadPoolExecutor
+from textwrap import dedent
+
+from bionic.deps.optdep import import_optional_dependency
+
+import hashlib
+import sys
+import tempfile
+import re
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+_cached_docker_module = None
+_cached_docker_client = None
+
+
+def get_docker_module():
+    global _cached_docker_module
+
+    if _cached_docker_module is None:
+        _cached_docker_module = import_optional_dependency(
+            "docker", purpose="Build Docker images"
+        )
+
+    return _cached_docker_module
+
+
+def get_docker_client():
+    global _cached_docker_client
+
+    if _cached_docker_client is None:
+        docker = get_docker_module()
+        logger.info("Initializing Docker client ...")
+        _cached_docker_client = docker.from_env()
+
+    return _cached_docker_client
+
+
+def fix_pip_requirements(pip_requirements: str) -> str:
+    # Pip freeze may contain entries with editable installs pointing to remote
+    # git repositories. This can happen when doing Bionic development. Docker
+    # service is not able to access repositories using the git+git protocol.
+    # Hence, any entries containing git+git is converted to use git+https.
+    #
+    # Example entry:
+    # -e git+git@github.com:square/bionic.git@f13f5405e928d92b553d2cbee41084eecccf7de3#egg=bionic
+    #
+    # Converted entry:
+    # -e git+https://github.com/square/bionic.git@f13f5405e928d92b553d2cbee41084eecccf7de3#egg=bionic
+
+    def fix_line(line: str) -> str:
+        if line.startswith("-e git+git"):
```

Comment:
In addition to `git`, `pip freeze` lines can also contain other VCS like Mercurial. I think it would be complicated and require a lot of effort to handle every possible combination. Only `git` is handled for now since Bionic is hosted in github.

Maybe we can add a warning whenever a line using VCS is detected?